### PR TITLE
Migrate BodyPartParser to GraphStage #20288.

### PR DESCRIPTION
**Pull request by danxmoran**
Ref #20288.

@johanandren here's BodyPartParser from #20327.  I'll open a follow-up with NaiveTake and ByteStringBatcher once this merges, to avoid adding noise with duplicate changes.

### CODE REVIEW

**Comment by johanandren** [comment]
OK TO TEST

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala [hunk 1]
```diff
@@ -57,202 +60,216 @@ private[http] final class BodyPartParser(defaultContentType: ContentType,
...
-          throw new IllegalStateException("unexpected NotEnoughDataException", NotEnoughDataException)
+  val in = Inlet[ByteString]("BodyPartParser.in")
+  val out = Outlet[BodyPartParser.Output]("BodyPartParser.out")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(attributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      private var output = collection.immutable.Queue.empty[Output]
+      // FIXME this probably is too wasteful
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: The fixme was about the queue before

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala [hunk 2]
```diff
@@ -57,202 +60,216 @@ private[http] final class BodyPartParser(defaultContentType: ContentType,
...
-  def parsePreamble(input: ByteString, offset: Int): StateResult =
-    try {
-      @tailrec def rec(index: Int): StateResult = {
-        val needleEnd = boyerMoore.nextIndex(input, index) + needle.length
-        if (crlf(input, needleEnd)) parseHeaderLines(input, needleEnd + 2)
-        else if (doubleDash(input, needleEnd)) terminate()
-        else rec(needleEnd)
+      override def onPull(): Unit = {
+        if (output.nonEmpty)
+          push(out, dequeue())
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: This formatting seems a bit off with the linefeed after if like this, did scalariform do it? (if not remove linefeed!)

File: akka-stream-tests/src/test/scala/akka/stream/impl/fusing/IteratorInterpreterSpec.scala [hunk 3]
```diff
@@ -8,21 +8,21 @@ import akka.util.ByteString
 import akka.stream.stage._
 import akka.stream.Supervision
 
-class IteratorInterpreterSpec extends AkkaSpec {
+class IteratorInterpreterSpec extends AkkaSpec with GraphInterpreterSpecKit {
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/IteratorInterpreterSpec.scala]: this doesn't belong in this PR?

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala [hunk 4]
```diff
@@ -57,202 +60,216 @@ private[http] final class BodyPartParser(defaultContentType: ContentType,
...
-    done()
-  }
+      def fail(info: ErrorInfo): StateResult = {
+        emit(ParseError(info))
+        terminate()
+      }
 
-  def done(): StateResult = null // StateResult is a phantom type
+      def terminate(): StateResult = {
+        terminated = true
```
Comment (viktorklang) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: I'm not 100 convinced of the name `terminated`. Isn't it more `shouldTerminate`?

File: akka-stream-tests/src/test/scala/akka/stream/impl/fusing/IteratorInterpreterSpec.scala (hunk 3 above)
Comment (danxmoran) [inline akka-stream-tests/src/test/scala/akka/stream/impl/fusing/IteratorInterpreterSpec.scala]: (As far as I know) I need it in order to call `.toGS` on the non-GraphStage-stages below.

... And I need to do that because I changed the `IteratorInterpreter` to take a `Seq[GraphStageWithMaterializedValue[FlowShape[_, _], Any]]` instead of a `Seq[PushPullStage[_, _]]`, because the only usage of `IteratorInterpreter` I could find was [with a `BodyPartParser`](https://github.com/akka/akka/blob/bd8fcc9d9aeafc2c49745e7519d13b3e0abc18e6/akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/MultipartUnmarshallers.scala#L84).

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala (hunk 4 above)
Comment (danxmoran) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: Yes, it does seem so.  Following that line of reasoning, should the method also be renamed from `terminate()` to something like `setShouldTerminate()`?

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala (hunk 2 above)
Comment (danxmoran) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: Not scalariform, just me mixing up style preferences between akka and work :) will fix

**Comment by patriknw** [comment]
LGTM

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala [hunk 5]
```diff
@@ -57,202 +60,216 @@ private[http] final class BodyPartParser(defaultContentType: ContentType,
...
-    }
 
-  @tailrec def parseHeaderLines(input: ByteString, lineStart: Int, headers: ListBuffer[HttpHeader] = ListBuffer[HttpHeader](),
-                                headerCount: Int = 0, cth: Option[`Content-Type`] = None): StateResult = {
-    def contentType =
-      cth match {
-        case Some(x) ⇒ x.contentType
-        case None    ⇒ defaultContentType
+      override def onUpstreamFinish(): Unit = {
+        if (isAvailable(out)) onPull()
```
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: Different behavior from the original stage which did just absorb termination? We are not sure there has been a pull from downstream as upstream finishes, so if output is `nonEmpty` this would lead to trying to push to non-ready downstream.

Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: Ah, sorry I'm not reading right today, `isAvailable`, but anyways different from the original stage.

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala [hunk 6]
```diff
@@ -57,202 +60,216 @@ private[http] final class BodyPartParser(defaultContentType: ContentType,
...
-          // we are missing a try/catch{continue} wrapper somewhere
-          throw new IllegalStateException("unexpected NotEnoughDataException", NotEnoughDataException)
+  val in = Inlet[ByteString]("BodyPartParser.in")
+  val out = Outlet[BodyPartParser.Output]("BodyPartParser.out")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogic(attributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) with InHandler with OutHandler {
+      private var output = collection.immutable.Queue.empty[Output]
```
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: We should replace this queue with something more efficient. Is there a reason here not using a mutable queue?

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala (hunk 4 above)
Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: yep, sounds like a plan!

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala (hunk 6 above)
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: I attempted that at some point btw and the results were surprising so this'll need a bit more investigation than it may seem at first: https://github.com/akka/akka/issues/18994****

Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: ok

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala (hunk 5 above)
Comment (danxmoran) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: Yes, I changed it according to the migration guide here: http://doc.akka.io/docs/akka-stream-and-http-experimental/2.0.4/scala/migration-guide-1.0-2.x-scala.html#AsyncStage_has_been_replaced_by_GraphStage

Am I misunderstanding the meaning of `<call the onPull() handler>`?

Comment (drewhk) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: the above is a one-to-one port of absorbTermination logic. We rarely do this style of encoding the logic, but for a port I think this is fine.

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala (hunk 6 above)
Comment (danxmoran) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: Should I leave replacing the queue to the other issue, then?

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala (hunk 5 above)
Comment (johanandren) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: Ah, sorry, my bad for not recognizing.

File: akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala (hunk 6 above)
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: I vote for doing it separately as it needs quite some benchmarking to be sure about the change.
If you'd like to puck that up go for it through, however we'll need benchmarks and profiling data :)

Comment (danxmoran) [inline akka-http-core/src/main/scala/akka/http/impl/engine/parsing/BodyPartParser.scala]: Sounds like a follow-up I'd be interested in trying out :) but I'll table it until I finish porting all the stages I've volunteered for.  In the meantime, I'll make sure the `FIXME` is in the right place.

**Comment by danxmoran** [comment]
Ok, I think I've responded to all comments (either via code change or follow-up comment).

**Comment by johanandren** [comment]
LGTM, thanks @danxmoran !
